state/relationunit.go

```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
-		return err
-	} else if count == 0 {
-		ops = append(ops, createSettingsOp(settingsC, ruKey, settings))
-	} else {
-		var rop txn.Op
-		rop, settingsChanged, err = replaceSettingsOp(ru.st.db(), settingsC, ruKey, settings)
-		if err != nil {
-			return err
-		}
-		ops = append(ops, rop)
-	}
 
-	// * Create the scope doc.
-	ops = append(ops, txn.Op{
-		C:      relationScopesC,
-		Id:     ruKey,
-		Assert: txn.DocMissing,
-		Insert: relationScopeDoc{
-			Key: ruKey,
-		},
-	})
+	var settingsChanged func() (bool, error)
 
-	// * If the unit should have a subordinate, and does not, create it.
 	var existingSubName string
-	if subOps, subName, err := ru.subordinateOps(); err != nil {
-		return err
-	} else {
-		existingSubName = subName
-		ops = append(ops, subOps...)
-	}
 
-	// Now run the complete transaction, or figure out why we can't.
-	if err := ru.st.db().RunTransaction(ops); err != txn.ErrAborted {
-		return err
-	}
-	if count, err := relationScopes.FindId(ruKey).Count(); err != nil {
-		return err
-	} else if count != 0 {
-		// The scope document exists, so we're actually already in scope.
-		return nil
-	}
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		// Before retrying the transaction, check the following
+		// assertions:
+		if attempt > 0 {
+			if count, err := relationScopes.FindId(ruKey).Count(); err != nil {
+				return nil, err
+			} else if count != 0 {
+				// The scope document exists, so we're actually already in scope.
+				return nil, nil
+			}
 
-	// The relation or unit might no longer be Alive. (Note that there is no
-	// need for additional checks if we're trying to create a subordinate
-	// unit: this could fail due to the subordinate applications not being Alive,
-	// but this case will always be caught by the check for the relation's
-	// life (because a relation cannot be Alive if its applications are not).)
-	relations, rCloser := db.GetCollection(relationsC)
-	defer rCloser()
-	if alive, err := isAliveWithSession(relations, relationDocID); err != nil {
-		return err
-	} else if !alive {
-		return stateerrors.ErrCannotEnterScope
-	}
-	if ru.isLocalUnit {
-		units, uCloser := db.GetCollection(unitsC)
-		defer uCloser()
-		if alive, err := isAliveWithSession(units, ru.unitName); err != nil {
-			return err
-		} else if !alive {
-			return stateerrors.ErrCannotEnterScope
+			// The relation or unit might no longer be Alive. (Note that there is no
+			// need for additional checks if we're trying to create a subordinate
+			// unit: this could fail due to the subordinate applications not being Alive,
+			// but this case will always be caught by the check for the relation's
+			// life (because a relation cannot be Alive if its applications are not).)
+			relations, rCloser := db.GetCollection(relationsC)
+			defer rCloser()
+			if alive, err := isAliveWithSession(relations, relationDocID); err != nil {
+				return nil, err
+			} else if !alive {
+				return nil, stateerrors.ErrCannotEnterScope
```

Comment:
Do we want something like:
errors.Annotate(stateerrors.ErrCannotEnterScope, "relation is no longer alive")

Reply:
I changed the enter scope errors a little bit, adding annotations on most of them